google-ads-examples/src/main/java/com/google/ads/googleads/examples/reporting/ParallelReportDownload.java

```diff
@@ -0,0 +1,253 @@
+   * Runs the example. This method may block for up to MAX_WAIT_TIME_SECONDS while waiting for
+   * reports to complete.
+   *
+   * @param googleAdsClient the client library instance for API access.
+   * @param customerIds the customer IDs to run against.
+   */
+  private void runExample(GoogleAdsClient googleAdsClient, List<Long> customerIds) {
+    // Creates a single client which can be shared by all threads.
+    // gRPC handles multiplexing parallel requests to the underlying API connection.
+    try (GoogleAdsServiceClient serviceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // IMPORTANT: You should avoid hitting the same customer ID in parallel. There are rate limits
+      // at the customer ID level which are much stricter than limits at the developer token level.
+      // Hitting these limits frequently enough will significantly reduce throughput as the client
+      // library will automatically retry with exponential back-off before failing the request.
+      for (String gaqlQuery : GAQL_QUERY_STRINGS) {
+        // Uses a list of futures to make sure that we wait for this report to complete on all
+        // customer IDs before proceeding. The Future data type here is just for demonstration.
+        List<ListenableFuture<ReportSummary>> futures = new ArrayList<>();
+
+        // Calls the API for each customer ID on the current report.
+        for (Long customerId : customerIds) {
+          // Uses the gRPC asynchronous API to download the reports in parallel. This saves having
+          // to create/manage our own thread pool.
+          ResponseCountingObserver responseObserver = new ResponseCountingObserver(customerId);
+
+          // Starts the report download in a background thread.
+          serviceClient
+              .searchStreamCallable()
+              .call(
+                  SearchGoogleAdsStreamRequest.newBuilder()
+                      .setCustomerId(customerId.toString())
+                      .setQuery(gaqlQuery)
+                      .build(),
+                  responseObserver);
+
+          // Stores a future to retrieve the results.
+          futures.add(responseObserver.asFuture());
+        }
+
+        // Waits for all pending requests to the current set of customer IDs to complete.
+        //
+        // This is a naive implementation for illustrative purposes. It is possible to optimize the
+        // utilization of each customer ID by providing a queue of work (or similar). However, this
+        // would complicate the example code and so is omitted here.
+        List<ReportSummary> results =
+            Futures.allAsList(futures)
+                .get(DEADLINE_MILLIS - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+        System.out.println("Report results for query: " + gaqlQuery);
+        results.forEach(System.out::println);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
Our style guide suggests that you should either:

1. Add `InterruptedException` to `throws` or
2. Call `interrupt()` as you've done here, then wrap in a custom exception

Since the wrapping here isn't adding any info or context, I'd go for the first option instead.

Reply:
Good point - done.